pp.c

```diff
@@ -3434,7 +3434,7 @@ PP(pp_substr)
             SV* repl_sv_copy = NULL;
 
             if (repl_need_utf8_upgrade) {
-                repl_sv_copy = newSVsv(repl_sv);
+                repl_sv_copy = newSVpvn(repl, repl_len);
```

Comment:
The old code could COW the copy, this will never do so.

That said, we know at this point we need to convert the byte per code point of the source to utf-8, and we don't actually need an SV to do the sv_insert_flags(), so we can just use bytes_to_utf8_free_me(), which won't even copy the string if the upgrade is trivial, ie. all of the code points are invariants.

Reply:
Cool, thanks. I'll take a look at `bytes_to_utf8_free_me()`.

Reply:
With the code in blead, `newSVsv(repl_sv)` could COW the copy, but `sv_utf8_upgrade` will un-COW it prior to doing the upgrade. So the conversion to UTF-8 will always happen on a buffer owned by just `repl_sv`.

I'm not quite sure what the change you're suggesting looks like. Surely I can't call `bytes_to_utf8_free_me` on a COWed buffer because of the state of the `IsCOW` flag and potentially the pointer to the buffer in other SV's sharing it?

Should I be doing `repl_sv_copy = newSVpvn(repl, repl_len);` followed by a call to `bytes_to_utf8_free_me`, and adding some code to fix up `repl_sv_copy`'s buffer pointer and IsCOW flag if required?

Reply:
Never mind, think I understand now!